### ISSUE TITLE
Infer `void` return type for function literal

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#32843]
[Edit Apr 20 2018 eernstg: I've added a test which directly tests the request made here, in [this comment](https://github.com/dart-lang/sdk/issues/32843#issuecomment-383039178). Note that CFE also infers return type `Null`, so we might need to make this a meta-bug and have separate bugs for the analyzer and CFE. Also note that we could easily extend this rule to say that also function literals with `return;` (but no `return e;`) should have inferred return type `void`.]

---

DDC appears to infer the return type `Null` for some function literals with a block body with no `return` statements in some cases, and it should infer `void`. Consider the following program:

```dart
void foo() {}

main() {
  (new Future.value(1)).then((_) {
    throw new Exception();
  }).catchError((_) => foo());
}
```

If the inferred type argument for `then` is `Null` then the dynamic requirement on the argument to `catchError` will be `FutureOr<Null> Function(...)` where `...` can stand for two distinct parameter type lists that we don't care about.

The type of `(_) => foo()` cannot be chosen such that it satisfies that constraint (the static type of the parameter is `Function`, but even if inference had "known" what to infer, that function literal could not have this type). Hence, execution stops with a dynamic error:

```
Type '(dynamic) => void' is not a subtype of type '(Object) => FutureOr<Null>'
```

With other tools than DDC, compilation as well as execution succeeds (presumably because `void` is inferred for `then`):

```
----- dart --checked: Test configuration: none_vm_release_x64_checked
[00:00 | 100% | +    1 | -    0]
----- dart2js --strong: Test configuration: dart2js_d8_release_x64_strong
[00:09 | 100% | +    1 | -    0]
----- dart2js --strong --checked: Test configuration: dart2js_d8_release_x64_checked_strong
[00:10 | 100% | +    1 | -    0]
----- dart2js_kernel --strong: Test configuration: dart2js_d8_release_x64_strong
[00:09 | 100% | +    1 | -    0]
----- dartk --strong: Test configuration: dartk_vm_release_x64_strong
[00:01 | 100% | +    1 | -    0]
----- dartk --strong --checked: Test configuration: dartk_vm_release_x64_checked_strong
[00:01 | 100% | +    1 | -    0]
----- dartkp --strong: Test configuration: dartkp_dart_precompiled_release_x64_strong
[00:10 | 100% | +    1 | -    0]
----- dartkp --strong --checked: Test configuration: dartkp_dart_precompiled_release_x64_checked_strong
[00:09 | 100% | +    1 | -    0]
----- dartdevk: Test configuration: dartdevk_chrome_release_x64_strong
[00:06 | 100% | +    1 | -    0]
----- dartdevk --checked: Test configuration: dartdevk_chrome_release_x64_checked_strong
[00:06 | 100% | +    1 | -    0]
```

but we have the following results for DDC:

```
----- dartdevc: Test configuration: dartdevc_chrome_release_x64_strong

FAILED: dartdevc-chrome release_x64 language_2/scratch_test
Expected: Pass
Actual: RuntimeError

--- Command "dartdevc" (took 01.000581s):
DART_CONFIGURATION=ReleaseX64 sdk/bin/dartdevc --dart-sdk-summary /usr/local/google/home/eernst/devel/dart/work/sdk/out/ReleaseX64/gen/utils/dartdevc/ddc_sdk.sum --ignore-unrecognized-flags --library-root /usr/local/google/home/eernst/devel/dart/work/sdk/tests/language_2 --module-root /usr/local/google/home/eernst/devel/dart/work/sdk/out/ReleaseX64/generated_compilations/dartdevc-strong --no-summarize --no-source-map -o /usr/local/google/home/eernst/devel/dart/work/sdk/out/ReleaseX64/generated_compilations/dartdevc-strong/tests_language_2_scratch_test/scratch_test.js /usr/local/google/home/eernst/devel/dart/work/sdk/tests/language_2/scratch_test.dart -s out/ReleaseX64/gen/utils/dartdevc/pkg/async_helper.sum=async_helper -s out/ReleaseX64/gen/utils/dartdevc/pkg/collection.sum=collection -s out/ReleaseX64/gen/utils/dartdevc/pkg/expect.sum=expect -s out/ReleaseX64/gen/utils/dartdevc/pkg/js.sum=js -s out/ReleaseX64/gen/utils/dartdevc/pkg/matcher.sum=matcher -s out/ReleaseX64/gen/utils/dartdevc/pkg/meta.sum=meta -s out/ReleaseX64/gen/utils/dartdevc/pkg/path.sum=path -s out/ReleaseX64/gen/utils/dartdevc/pkg/stack_trace.sum=stack_trace -s out/ReleaseX64/gen/utils/dartdevc/pkg/unittest.sum=unittest

exit code:
0

--- Command "chrome" (took 01.000063s):
/usr/local/google/home/eernst/devel/dart/work/sdk/tools/sdks/linux/dart-sdk/bin/dart tools/testing/dart/launch_browser.dart chrome http://127.0.0.1:43217/root_build/generated_tests/dartdevc-chrome-strong/tests_language_2_scratch_test/test.html?crossOriginPort=40425

Runtime window.onerror:
window.onerror called: 

http://127.0.0.1:43217/root_build/gen/utils/dartdevc/js/amd/dart_sdk.js:4976:
Uncaught Type '(dynamic) => void' is not a subtype of type '(Object) => FutureOr<Null>'

exit code:
0

--- To debug locally, run:
/usr/local/google/home/eernst/devel/dart/work/sdk/tools/sdks/linux/dart-sdk/bin/dart /usr/local/google/home/eernst/devel/dart/work/sdk/tools/testing/dart/http_server.dart -p 43217 -c 40425 --network 127.0.0.1 --build-directory=/usr/local/google/home/eernst/devel/dart/work/sdk/out/ReleaseX64/ --runtime=chrome --packages=/usr/local/google/home/eernst/devel/dart/work/sdk/.packages

--- Re-run this test:
python tools/test.py -m release -c dartdevc --strong --host-checked -t 960 language_2/scratch_test


[00:05 | 100% | +    0 | -    1]
Double reporting tests
  http://127.0.0.1:43217/root_build/generated_tests/dartdevc-chrome-strong/tests_language_2_scratch_test/test.html?crossOriginPort=40425
2018-04-11 11:56:40.299063 Warning: Double reporting tests:
2018-04-11 11:56:40.309946 Warning: http://127.0.0.1:43217/root_build/generated_tests/dartdevc-chrome-strong/tests_language_2_scratch_test/test.html?crossOriginPort=40425, output: 
2018-04-11 11:56:40.310025 Warning: [{"type":"debug","value":"test_controller.js started","timestamp":0},{"type":"debug","value":"Sending events to driver page (isFirstMessage = true, isStatusUpdate = true, isDone = false)","timestamp":0},{"type":"print","value":"dart-calling-main","timestamp":0.935},{"type":"print","value":"dart-main-done","timestamp":0.937},{"type":"window_onerror","value":"window.onerror called: \n\nhttp://127.0.0.1:43217/root_build/gen/utils/dartdevc/js/amd/dart_sdk.js:4976:\nUncaught Type '(dynamic) => void' is not a subtype of type '(Object) => FutureOr<Null>'\n\n","timestamp":0.952},{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":0.952},{"type":"message_received","value":"dart-calling-main","timestamp":0.953},{"type":"message_received","value":"dart-main-done","timestamp":0.953},{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":0.953},{"type":"dom","value":"<head>\n  <meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n  <meta name=\"dart.unittest\" content=\"full-stack-traces\">\n  <title>Test scratch_test</title>\n  <style>\n     .unittest-table { font-family:monospace; border:1px; }\n     .unittest-pass { background: #6b3;}\n     .unittest-fail { background: #d55;}\n     .unittest-error { background: #a11;}\n  </style>\n<script type=\"text/javascript\" charset=\"utf-8\" async=\"\" data-requirecontext=\"_\" data-requiremodule=\"scratch_test\" src=\"/root_dart/out/ReleaseX64/generated_compilations/dartdevc-strong/tests_language_2_scratch_test/scratch_test.js\"></script><script type=\"text/javascript\" charset=\"utf-8\" async=\"\" data-requirecontext=\"_\" data-requiremodule=\"dart_sdk\" src=\"/root_build/gen/utils/dartdevc/js/amd/dart_sdk.js\"></script><script type=\"text/javascript\" charset=\"utf-8\" async=\"\" data-requirecontext=\"_\" data-requiremodule=\"async_helper\" src=\"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/async_helper.js\"></script></head>\n<body>\n<h1>Running scratch_test</h1>\n<script type=\"text/javascript\" src=\"/root_dart/tools/testing/dart/test_controller.js\">\n</script>\n<script>\nvar require = {\n  baseUrl: \"/root_dart/out/ReleaseX64/generated_compilations/dartdevc-strong/tests_language_2_scratch_test\",\n  paths: {\n    \"dart_sdk\": \"/root_build/gen/utils/dartdevc/js/amd/dart_sdk\",\n    \"async_helper\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/async_helper\",\n    \"collection\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/collection\",\n    \"expect\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/expect\",\n    \"js\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/js\",\n    \"matcher\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/matcher\",\n    \"meta\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/meta\",\n    \"path\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/path\",\n    \"stack_trace\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/stack_trace\",\n    \"unittest\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/unittest\",\n  },\n  waitSeconds: 30,\n};\n\n// Don't try to bring up the debugger on a runtime error.\nwindow.ddcSettings = {\n  trapRuntimeErrors: false\n};\n</script>\n<script type=\"text/javascript\" src=\"/root_dart/third_party/requirejs/require.js\"></script>\n<script type=\"text/javascript\">\nrequirejs([\"scratch_test\", \"dart_sdk\", \"async_helper\"],\n    function(scratch_test, sdk, async_helper) {  \n  sdk.dart.ignoreWhitelistedErrors(false);\n  if (true) sdk.dart.setStartAsyncSynchronously();\n  // TODO(rnystrom): This uses DDC's forked version of async_helper. Unfork\n  // these packages when possible.\n  async_helper.async_helper.asyncTestInitialize(function() {});\n  sdk._isolate_helper.startRootIsolate(function() {}, []);\n  sdk._debugger.registerDevtoolsFormatter();\n\n  testErrorToStackTrace = function(error) {\n    var stackTrace = sdk.dart.stackTrace(error).toString();\n    \n    var lines = stackTrace.split(\"\\n\");\n    \n    // Remove the first line, which is just \"Error\".\n    lines = lines.slice(1);\n\n    // Strip off all of the lines for the bowels of the test runner.\n    for (var i = 0; i < lines.length; i++) {\n      if (lines[i].indexOf(\"dartMainRunner\") != -1) {\n        lines = lines.slice(0, i);\n        break;\n      }\n    }\n    \n    // TODO(rnystrom): It would be nice to shorten the URLs of the remaining\n    // lines too.\n    return lines.join(\"\\n\");\n  };\n  \n  dartMainRunner(scratch_test.scratch_test.main);\n});\n</script>\n\n\n</body>","timestamp":0.953}]
2018-04-11 11:56:40.310071 Warning: 
2018-04-11 11:56:40.310097 Warning: 

[00:06 | 100% | +    0 | -    1]
----- dartdevc --checked: Test configuration: dartdevc_chrome_release_x64_checked_strong

FAILED: dartdevc-chrome-checked release_x64 language_2/scratch_test
Expected: Pass
Actual: RuntimeError

--- Command "dartdevc" (took 01.000388s):
DART_CONFIGURATION=ReleaseX64 sdk/bin/dartdevc --dart-sdk-summary /usr/local/google/home/eernst/devel/dart/work/sdk/out/ReleaseX64/gen/utils/dartdevc/ddc_sdk.sum --ignore-unrecognized-flags --library-root /usr/local/google/home/eernst/devel/dart/work/sdk/tests/language_2 --module-root /usr/local/google/home/eernst/devel/dart/work/sdk/out/ReleaseX64/generated_compilations/dartdevc-checked-strong --no-summarize --no-source-map -o /usr/local/google/home/eernst/devel/dart/work/sdk/out/ReleaseX64/generated_compilations/dartdevc-checked-strong/tests_language_2_scratch_test/scratch_test.js /usr/local/google/home/eernst/devel/dart/work/sdk/tests/language_2/scratch_test.dart -s out/ReleaseX64/gen/utils/dartdevc/pkg/async_helper.sum=async_helper -s out/ReleaseX64/gen/utils/dartdevc/pkg/collection.sum=collection -s out/ReleaseX64/gen/utils/dartdevc/pkg/expect.sum=expect -s out/ReleaseX64/gen/utils/dartdevc/pkg/js.sum=js -s out/ReleaseX64/gen/utils/dartdevc/pkg/matcher.sum=matcher -s out/ReleaseX64/gen/utils/dartdevc/pkg/meta.sum=meta -s out/ReleaseX64/gen/utils/dartdevc/pkg/path.sum=path -s out/ReleaseX64/gen/utils/dartdevc/pkg/stack_trace.sum=stack_trace -s out/ReleaseX64/gen/utils/dartdevc/pkg/unittest.sum=unittest

exit code:
0

--- Command "chrome" (took 01.000053s):
/usr/local/google/home/eernst/devel/dart/work/sdk/tools/sdks/linux/dart-sdk/bin/dart tools/testing/dart/launch_browser.dart chrome http://127.0.0.1:40113/root_build/generated_tests/dartdevc-chrome-checked-strong/tests_language_2_scratch_test/test.html?crossOriginPort=35369

Runtime window.onerror:
window.onerror called: 

http://127.0.0.1:40113/root_build/gen/utils/dartdevc/js/amd/dart_sdk.js:4976:
Uncaught Type '(dynamic) => void' is not a subtype of type '(Object) => FutureOr<Null>'

exit code:
0

--- To debug locally, run:
/usr/local/google/home/eernst/devel/dart/work/sdk/tools/sdks/linux/dart-sdk/bin/dart /usr/local/google/home/eernst/devel/dart/work/sdk/tools/testing/dart/http_server.dart -p 40113 -c 35369 --network 127.0.0.1 --build-directory=/usr/local/google/home/eernst/devel/dart/work/sdk/out/ReleaseX64/ --runtime=chrome --packages=/usr/local/google/home/eernst/devel/dart/work/sdk/.packages

--- Re-run this test:
python tools/test.py -m release -c dartdevc --checked --strong --host-checked -t 960 language_2/scratch_test


[00:05 | 100% | +    0 | -    1]
Double reporting tests
  http://127.0.0.1:40113/root_build/generated_tests/dartdevc-chrome-checked-strong/tests_language_2_scratch_test/test.html?crossOriginPort=35369
2018-04-11 11:56:46.182691 Warning: Double reporting tests:
2018-04-11 11:56:46.190594 Warning: http://127.0.0.1:40113/root_build/generated_tests/dartdevc-chrome-checked-strong/tests_language_2_scratch_test/test.html?crossOriginPort=35369, output: 
2018-04-11 11:56:46.190670 Warning: [{"type":"debug","value":"test_controller.js started","timestamp":0},{"type":"debug","value":"Sending events to driver page (isFirstMessage = true, isStatusUpdate = true, isDone = false)","timestamp":0},{"type":"print","value":"dart-calling-main","timestamp":0.952},{"type":"print","value":"dart-main-done","timestamp":0.955},{"type":"window_onerror","value":"window.onerror called: \n\nhttp://127.0.0.1:40113/root_build/gen/utils/dartdevc/js/amd/dart_sdk.js:4976:\nUncaught Type '(dynamic) => void' is not a subtype of type '(Object) => FutureOr<Null>'\n\n","timestamp":0.97},{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":0.97},{"type":"message_received","value":"dart-calling-main","timestamp":0.971},{"type":"message_received","value":"dart-main-done","timestamp":0.972},{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":0.972},{"type":"dom","value":"<head>\n  <meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n  <meta name=\"dart.unittest\" content=\"full-stack-traces\">\n  <title>Test scratch_test</title>\n  <style>\n     .unittest-table { font-family:monospace; border:1px; }\n     .unittest-pass { background: #6b3;}\n     .unittest-fail { background: #d55;}\n     .unittest-error { background: #a11;}\n  </style>\n<script type=\"text/javascript\" charset=\"utf-8\" async=\"\" data-requirecontext=\"_\" data-requiremodule=\"scratch_test\" src=\"/root_dart/out/ReleaseX64/generated_compilations/dartdevc-checked-strong/tests_language_2_scratch_test/scratch_test.js\"></script><script type=\"text/javascript\" charset=\"utf-8\" async=\"\" data-requirecontext=\"_\" data-requiremodule=\"dart_sdk\" src=\"/root_build/gen/utils/dartdevc/js/amd/dart_sdk.js\"></script><script type=\"text/javascript\" charset=\"utf-8\" async=\"\" data-requirecontext=\"_\" data-requiremodule=\"async_helper\" src=\"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/async_helper.js\"></script></head>\n<body>\n<h1>Running scratch_test</h1>\n<script type=\"text/javascript\" src=\"/root_dart/tools/testing/dart/test_controller.js\">\n</script>\n<script>\nvar require = {\n  baseUrl: \"/root_dart/out/ReleaseX64/generated_compilations/dartdevc-checked-strong/tests_language_2_scratch_test\",\n  paths: {\n    \"dart_sdk\": \"/root_build/gen/utils/dartdevc/js/amd/dart_sdk\",\n    \"async_helper\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/async_helper\",\n    \"collection\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/collection\",\n    \"expect\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/expect\",\n    \"js\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/js\",\n    \"matcher\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/matcher\",\n    \"meta\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/meta\",\n    \"path\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/path\",\n    \"stack_trace\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/stack_trace\",\n    \"unittest\": \"/root_dart/out/ReleaseX64/gen/utils/dartdevc/pkg/unittest\",\n  },\n  waitSeconds: 30,\n};\n\n// Don't try to bring up the debugger on a runtime error.\nwindow.ddcSettings = {\n  trapRuntimeErrors: false\n};\n</script>\n<script type=\"text/javascript\" src=\"/root_dart/third_party/requirejs/require.js\"></script>\n<script type=\"text/javascript\">\nrequirejs([\"scratch_test\", \"dart_sdk\", \"async_helper\"],\n    function(scratch_test, sdk, async_helper) {  \n  sdk.dart.ignoreWhitelistedErrors(false);\n  if (true) sdk.dart.setStartAsyncSynchronously();\n  // TODO(rnystrom): This uses DDC's forked version of async_helper. Unfork\n  // these packages when possible.\n  async_helper.async_helper.asyncTestInitialize(function() {});\n  sdk._isolate_helper.startRootIsolate(function() {}, []);\n  sdk._debugger.registerDevtoolsFormatter();\n\n  testErrorToStackTrace = function(error) {\n    var stackTrace = sdk.dart.stackTrace(error).toString();\n    \n    var lines = stackTrace.split(\"\\n\");\n    \n    // Remove the first line, which is just \"Error\".\n    lines = lines.slice(1);\n\n    // Strip off all of the lines for the bowels of the test runner.\n    for (var i = 0; i < lines.length; i++) {\n      if (lines[i].indexOf(\"dartMainRunner\") != -1) {\n        lines = lines.slice(0, i);\n        break;\n      }\n    }\n    \n    // TODO(rnystrom): It would be nice to shorten the URLs of the remaining\n    // lines too.\n    return lines.join(\"\\n\");\n  };\n  \n  dartMainRunner(scratch_test.scratch_test.main);\n});\n</script>\n\n\n</body>","timestamp":0.972}]
2018-04-11 11:56:46.190710 Warning: 
2018-04-11 11:56:46.192027 Warning: 

[00:05 | 100% | +    0 | -    1]
```

The SDK commit version used for these test runs was 678691334fdaca5c4878e062b33736b13f7a3ecc, with date 2018, Apr 11, 04:10:25.

**Comment by jmesserly** [comment]
This is an Analyzer issue. IDK how much we're prioritizing fixing Analyzer issues since it will be replaced with CFE soon (thoughts @leafpetersen ?)

**Comment by leafpetersen** [comment]
I really want to fix this one.  Have we verified that the CFE does the right thing?  I have a slightly out of date CL implementing this, and making the necessary changes to internal dart code, so if the CFE is doing it right we're safe to land it.  Otherwise I need to do some work on flutter code to make sure it's compatible.  We also need to fix an issue around `return;` in async functions as part of this.

**Comment by jmesserly** [comment]
based on @eernstg comment that it works everywhere but DDC, it sounds like CFE is working

**Comment by eernstg** [comment]
Ah, that's probably because my mental model concerning the relevant cases is too coarse-grained.

Since every case except the last one leaves the type of the function literal entirely unconstrained (there is no type-from-context), the inferred type would be the "natural" type for the given function literal; and that turns out to yield a return type of `Null` (async: `Future<Null>`) in all the cases where there is no `return e;` (which also rules out `=>`), but we may reach the end and/or execute `return;`. As I mentioned in [Buganizer](https://buganizer.corp.google.com/issues/62238091#comment22), this seems to be the optimal choice, in that it is sound and it allows for the greatest possible set of usages of the function.

However, we did note in many discussions that it is also dangerous, because such a function would pass for a `WhateverYouWant Function()`, and it would be better to reject the function than having an unintended `null` when it's finally called. So that's an argument in favor of giving all these functions a return type of `void`: "The developer did not bother to deliver a value, so we shouldn't consider the returned value to be valid, and the return type `void` says exactly that".

My takeaway from Leaf's remark 'We should infer void' [here](https://buganizer.corp.google.com/issues/62238091#comment23) was that the latter argument had won. We could still infer other return types in cases where `void` makes it impossible to type the enclosing expression&mdash;but that pretty much defies the whole point, because we would then still get those unintended-`null` bugs when the function is called, and there's no warning or anything.

So I concluded that we would type `(...) { .../*no returns*/... }` and its ilk as `void Function(...)`, but obviously that's not true. (Maybe it should be true? I guess it would break all kinds of stuff ;-).

For last example, the type-from-context is `FutureOr<S> Function(int)` where `S` is unconstrained (and presumably turns out to be `dynamic`, such that `then` returns a `Future<dynamic>`). So, apparently, CFE uses the type-from-context, which is allowed because `null` is an acceptable returned value for a function of type `FutureOr<dynamic> Function(int)`, and DDC might be using the "natural" type `Null Function(int)`.

This means that (1) we're inferring the return type `void` in a surprisingly small number of situations for a function literal, and (2) only DDC uses the freedom to assign the "natural" type to a function literal, cf. [this issue comment](https://github.com/dart-lang/sdk/issues/31693#issuecomment-353048273).
